R/kill-parallel.R

```diff
@@ -0,0 +1,33 @@
+#' Kill processes with grace period
+#'
+#' On Unixes, `ps_kill_parallel()` sends a `SIGTERM` to all processes
+#' in `ps`. If the processes are not killed after a grace period
+#' defined by `grace`, a `SIGKILL` is issued. On Windows, the
+#' processes are abruptly terminated without grace period.
+#'
+#' Direct subprocesses are polled for termination events, which means
+#' that `ps_kill_parallel()` will return earlier than the grace period
+#' as soon as all `SIGTERM` signals are successful. However if `ps`
+#' includes subsubprocesses, those are periodically polled for
+#' termination as ps doesn't receive termination events for those. In
+#' this case, `ps_kill_parallel()` might return a little later.
+#'
+#' @param ps List of [process handles](ps_handle).
+#' @param grace Grace period between `SIGTERM` and `SIGKILL` in seconds.
+#' @export
+ps_kill_parallel <- function(ps, grace = 0.2) {
+  stopifnot(
+    is_handle_list(ps),
+    is_scalar_numeric(grace)
+  )
+
+  if (.Platform$OS == "windows") {
+    for (p in ps) {
+      ps_kill(p)
+    }
+  } else {
+    .Call(ps__kill_parallel, ps, grace)
+  }
+
+  NULL
```

Comment:
Here I'm returning `NULL` because it's not clear to me that we have any relevant information to return to the caller. We could return a logical vector that indicates whether we sent signals for the process or we refrained because the creation time didn't match, but this makes things more complicated and provides information that maybe shouldn't be relied on.